README.md
# @mconnect/mcauditlog

- Transactions / audit log for data CRUD functions and access tracking
- Includes audit/transaction logs to postgresDB | mongoDB
- See the test folder for different test cases / scenarios and usage

## Installation
```js
npm i @mconnect/mcauditlog
```


HISTORY.md
# Historical logs

## Version 0.2.5

- Release date: Jun 06, 2021
- updated mcresponse package to v0.4.1 & mcdb to v0.2.7

## Version 0.2.4

- Release date: Jun 06, 2021
- updated mcresponse package to v0.4.1 & mcdb to v0.2.6


## Version 0.2.3

- Release date: Jan 25, 2021
- updated mcdb package to v0.2.5

## Version 0.2.2

- Release date: Jan 25, 2021
- updated mcdb package

## Version 0.2.1

- Release date: Dec 14, 2020
- patched with mongo and validator-types

## Version 0.2.0

- Release date: Dec 14, 2020
- npm/github publication: refactor, test, build and release

## Version 0.1.0
- Release date: July 16, 2020 
- Includes audit/transaction logs to mongoDB collection
- See the test folder for different test cases / scenarios and usage


test/model.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-07-16
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: mc-central-ts: audit-log model (Audit => audits)
 */

import { DbConnectType } from "@mconnect/mcdb";
import { DataTypes, Model } from "sequelize";

export async function auditModel(dbc: DbConnectType) {
    class Audit extends Model {
    }

    Audit.init({
        id              : {
            type        : DataTypes.UUID,
            primaryKey  : true,
            allowNull   : false,
            defaultValue: DataTypes.UUIDV4,
            validate    : {
                isUUID: 4
            },
        },
        collName        : {
            type     : DataTypes.STRING,
            allowNull: false
        },
        collDocuments   : {
            type: DataTypes.JSONB,
        },
        newCollDocuments: {
            type: DataTypes.JSONB,
        },
        logType         : {
            type     : DataTypes.STRING,
            allowNull: false,
        },
        logBy           : {
            type    : DataTypes.UUID,
            validate: {
                isUUID: 4
            },
        },
        logAt           : {
            type        : DataTypes.DATE,
            allowNull   : false,
            defaultValue: DataTypes.NOW,
        },
    }, {
        sequelize  : dbc,
        tableName  : "audits",
        modelName  : "Audit",
        timestamps : false,
        underscored: true,
    });

    // Sync the table
    await Audit.sync({alter: true});

    return Audit;
}


test/auditLog.test.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-07-15
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: mc-central-ts: audit-log testing (sequelize)
 */

import { mcTest, assertEquals, postTestResult } from "@mconnect/mctest";
import { newAuditLog } from "../src";
import { newDbConnect } from "@mconnect/mcdb";

const dbs = {
    postgres: {
        location: process.env.PGDB_URL || "postgres://localhost:5432/mc-dev",
        host    : process.env.PGDB_HOST || "localhost",
        username: process.env.PGDB_USER || "postgres",
        password: process.env.PGDB_PWD || "ab12testing",
        database: process.env.PDDB_DBNAME || "mccentral",
        port    : process.env.PGDB_PORT || 5432,
    }
}

import { auditModel } from "./model";

// test data
const
    collName = "services",
    userId = "faea411c-e82d-454f-8ee7-574c4e753d06";

const collDocuments: any = {
    name    : "Abi",
    desc    : "Testing only",
    url     : "localhost:9000",
    priority: 1,
    cost    : 1000.00
};
const newCollDocuments: any = {
    name    : "Abi Akindele",
    desc    : "Testing only - updated",
    url     : "localhost:9900",
    priority: 1,
    cost    : 2000.00
};

(async () => {
    // pre-testing setup
    const dbc = newDbConnect(dbs.postgres, "postgres");
    const auditDb = await dbc?.openDb();
    // model
    const Audit = await auditModel(auditDb);
    const auditLog = newAuditLog(Audit, {auditTable: "auditors"});

    // perform audit-log test tasks
    await mcTest({
        name    : "should connect and return an instance object: ",
        testFunc: async () => {
            assertEquals(Object.keys(auditLog).length > 0, true);
        },
    });

    await mcTest({
        name    : "should store create-transaction log and return success: ",
        testFunc: async () => {
            let res = await auditLog.createLog(collName, collDocuments, userId);
            // console.log("create-log-result: ", res);
            assertEquals(res.code, "success");
        },
    });

    await mcTest({
        name    : "should store update-transaction log and return success: ",
        testFunc: async () => {
            let res = await auditLog.updateLog(collName, collDocuments, newCollDocuments, userId);
            assertEquals(res.code, "success");
        },
    });

    await mcTest({
        name    : "should store read-transaction log and return success: ",
        testFunc: async () => {
            const recParams = {keywords: ["lagos", "nigeria", "ghana", "accra"]};
            let res = await auditLog.readLog(collName, recParams, userId);
            assertEquals(res.code, "success");
        },
    });

    await mcTest({
        name    : "should store delete-transaction log and return success: ",
        testFunc: async () => {
            let res = await auditLog.deleteLog(collName, collDocuments, userId);
            assertEquals(res.code, "success");
        },
    });

    await mcTest({
        name    : "should store login-transaction log and return success: ",
        testFunc: async () => {
            let res = await auditLog.loginLog(collDocuments);
            assertEquals(res.code, "success");
        },
    });

    await mcTest({
        name    : "should store logout-transaction log and return success: ",
        testFunc: async () => {
            let res = await auditLog.logoutLog(collDocuments, userId);
            assertEquals(res.code, "success");
        },
    });

    await mcTest({
        name    : "should return paramsError for incomplete/undefined inputs: ",
        testFunc: async () => {
            let res = await auditLog.createLog(collName, collDocuments, "");
            assertEquals(res.code, "logError");
            assertEquals(res.message.includes("userId is required"), true);
        },
    });

    await mcTest({
        name    : "generic-audit-log/create: should store transaction log and return success: ",
        testFunc: async () => {
            let res = await auditLog.auditLog("Create", userId, {collName, collDocuments});
            assertEquals(res.code, "success");
        },
    });

    await mcTest({
        name    : "generic-audit-log/update: should store transaction log and return success: ",
        testFunc: async () => {
            let res = await auditLog.auditLog("Update", userId, {collName, collDocuments, newCollDocuments});
            assertEquals(res.code, "success");
        },
    });

    // post testing report
    await postTestResult();

    // close resources / avoid memory leak
    await dbc?.closeDb();
})();


test/auditLogMongo.test.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-07-17
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: mc-central-ts: audit-log testing (mongodb)
 */

import { mcTest, assertEquals, postTestResult } from "@mconnect/mctest";
import { newAuditLogMongo } from "../src";
import { newDbMongo } from "@mconnect/mcdb";

// test data
const dbs = {
    mongodb: {
        location: process.env.MONGODB_URL || "mongodb://localhost:27017/mc-central",
        host    : process.env.MONGODB_HOST || "localhost",
        username: process.env.MONGODB_USER || "abbeymart",
        password: process.env.MONGODB_PWD || "ab12testing",
        database: process.env.MONGODB_DBNAME || "mc-central",
        port    : process.env.MONGODB_PORT || 27017,
    }
}

const
    collName = "services",
    userId = "abbeycityunited2020";

const collDocuments: any = {
    name    : "Abi",
    desc    : "Testing only",
    url     : "localhost:9000",
    priority: 1,
    cost    : 1000.00
};
const newCollDocuments: any = {
    name    : "Abi Akindele",
    desc    : "Testing only - updated",
    url     : "localhost:9900",
    priority: 1,
    cost    : 2000.00
};

(async () => {
    // pre-testing setup
    let dbc = await newDbMongo(dbs.mongodb, {checkAccess: false});
    let auditDb = await dbc.openDb("mccentral");

    const auditLog = newAuditLogMongo(auditDb, {auditColl: "audits"});

    // perform audit-log test tasks
    await mcTest({
        name    : "should connect and return an instance object: ",
        testFunc: async () => {
            assertEquals(Object.keys(auditLog).length > 0, true);
        },
    });

    await mcTest({
        name    : "should store create-transaction log and return success: ",
        testFunc: async () => {
            let res = await auditLog.createLog(collName, collDocuments, userId);
            // console.log("create-log-result-100: ", res);
            assertEquals(res.code, "success");
        },
    });

    await mcTest({
        name    : "should store update-transaction log and return success: ",
        testFunc: async () => {
            let res = await auditLog.updateLog(collName, collDocuments, newCollDocuments, userId);
            assertEquals(res.code, "success");
        },
    });

    await mcTest({
        name    : "should store read-transaction log and return success: ",
        testFunc: async () => {
            const recParams = {keywords: ["lagos", "nigeria", "ghana", "accra"]};
            let res = await auditLog.readLog(collName, recParams, userId);
            assertEquals(res.code, "success");
        },
    });

    await mcTest({
        name    : "should store delete-transaction log and return success: ",
        testFunc: async () => {
            let res = await auditLog.deleteLog(collName, collDocuments, userId);
            assertEquals(res.code, "success");
        },
    });

    await mcTest({
        name    : "should store login-transaction log and return success: ",
        testFunc: async () => {
            let res = await auditLog.loginLog(collDocuments);
            assertEquals(res.code, "success");
        },
    });

    await mcTest({
        name    : "should store logout-transaction log and return success: ",
        testFunc: async () => {
            let res = await auditLog.logoutLog(collDocuments, userId);
            assertEquals(res.code, "success");
        },
    });

    await mcTest({
        name    : "should return paramsError for incomplete/undefined inputs: ",
        testFunc: async () => {
            let res = await auditLog.createLog(collName, collDocuments, "");
            assertEquals(res.code, "logError");
            assertEquals(res.message.includes("userId is required"), true);
        },
    });

    await mcTest({
        name    : "generic-audit-log/create: should store transaction log and return success: ",
        testFunc: async () => {
            let res = await auditLog.auditLog("Create", userId, {collName, collDocuments});
            assertEquals(res.code, "success");
        },
    });

    await mcTest({
        name    : "generic-audit-log/update: should store transaction log and return success: ",
        testFunc: async () => {
            let res = await auditLog.auditLog("Update", userId, {collName, collDocuments, newCollDocuments});
            assertEquals(res.code, "success");
        },
    });


    // post testing report
    await postTestResult();

    // close resources / avoid memory leak
    await dbc?.closeDb();
})();


src/auditLogMongo.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-07-15
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: mc-central-ts: audit-log (mongodb) entry point | translog
 */

// Import required module/function(s)
import { MongoDbConnectType, checkDb } from "@mconnect/mcdb";
import { getResMessage, ResponseMessage } from "@mconnect/mcresponse";

//types
export interface AuditLogOptionsType {
    auditColl?: string;
    collName?: string;
    collDocuments?: any;
    newCollDocuments?: any;
    recordParams?: any;
    newRecordParams?: any;
}

export enum AuditLogTypes {
    CREATE = "create",
    UPDATE = "update",
    DELETE = "delete",
    REMOVE = "remove",
    GET = "get",
    READ = "read",
    LOGIN = "login",
    LOGOUT = "logout",
}

class AuditLogMongo {
    private readonly dbConnect: MongoDbConnectType;
    private readonly auditColl: string;

    constructor(auditDb: MongoDbConnectType, options?: AuditLogOptionsType) {
        this.dbConnect = auditDb;
        this.auditColl = options && options.auditColl ? options.auditColl : "audits";
    }

    async createLog(collName: string, collDocuments: any, userId: string): Promise<ResponseMessage> {
        const dbCheck = checkDb(this.dbConnect);
        if (dbCheck.code !== "success") {
            return dbCheck;
        }

        // Check/validate the attributes / parameters
        let errorMessage = "";
        if (!collName) {
            errorMessage = errorMessage ? errorMessage + " | Table or Collection name is required." :
                "Table or Collection name is required.";
        }
        if (!userId) {
            errorMessage = errorMessage ? errorMessage + " | userId is required." :
                "userId is required.";
        }
        if (!collDocuments) {
            errorMessage = errorMessage ? errorMessage + " | Created record(s) information is required." :
                "Created record(s) information is required.";
        }
        if (errorMessage) {
            console.log("error-message: ", errorMessage);
            return getResMessage("logError", {
                message: errorMessage,
            });
        }

        try {
            // insert audit record
            const coll = this.dbConnect.collection(this.auditColl);
            const result = await coll.insertOne({
                collName     : collName,
                collDocuments: collDocuments,
                logType      : "create",
                logBy        : userId,
                logAt        : new Date(),
            });

            if (result) {
                return getResMessage("success", {
                    value: result,
                });
            } else {
                return getResMessage("insertError", {
                    value  : result || "no-result",
                    message: "no response from the server",
                });
            }
        } catch (error) {
            console.log("Error saving create-audit record(s): ", error);
            return getResMessage("logError", {
                value  : error.message,
                message: "Error saving create-audit record(s): " + error.message,
            });
        }
    }

    async updateLog(collName: string, collDocuments: any, newCollDocuments: any, userId: string): Promise<ResponseMessage> {
        const dbCheck = checkDb(this.dbConnect);
        if (!dbCheck) {
            return dbCheck;
        }

        // Check/validate the attributes / parameters
        let errorMessage = "";
        if (!collName) {
            errorMessage = errorMessage ? errorMessage + " | Table or Collection name is required." :
                "Table or Collection name is required.";
        }
        if (!userId) {
            errorMessage = errorMessage ? errorMessage + " | userId is required." :
                "userId is required.";
        }
        if (!collDocuments) {
            errorMessage = errorMessage ? errorMessage + " | Current record(s) information is required." :
                "Current record(s) information is required.";
        }
        if (!newCollDocuments) {
            errorMessage = errorMessage ? errorMessage + " | Updated record(s) information is required." :
                "Updated record(s) information is required.";
        }
        if (errorMessage) {
            return getResMessage("insertError", {
                message: errorMessage,
            });
        }

        try {
            // insert audit record
            const coll = this.dbConnect.collection(this.auditColl);
            const result = await coll.insertOne({
                collName        : collName,
                collDocuments   : collDocuments,
                newCollDocuments: newCollDocuments,
                logType         : "update",
                logBy           : userId,
                logAt           : new Date(),
            });

            if (result) {
                return getResMessage("success", {
                    value: result,
                });
            } else {
                return getResMessage("insertError");
            }
        } catch (error) {
            console.error("Error saving update-audit record(s): ", error);
            return getResMessage("insertError", {
                message: "Error saving update-audit record(s): " + error.message,
            });
        }
    }

    async readLog(collName: string, collDocuments: any, userId: string = ""): Promise<ResponseMessage> {
        const dbCheck = checkDb(this.dbConnect);
        if (!dbCheck) {
            return dbCheck;
        }

        // validate params/values
        let errorMessage = "";
        if (!collName) {
            errorMessage = errorMessage ? errorMessage + " | Table or Collection name is required." :
                "Table or Collection name is required.";
        }
        if (!collDocuments) {
            errorMessage = errorMessage ? errorMessage + " | Search keywords or Read record(s) information is required." :
                "Search keywords or Read record(s) information is required.";
        }
        if (errorMessage) {
            return getResMessage("insertError", {
                message: errorMessage,
            });
        }

        try {
            // insert audit record
            const coll = this.dbConnect.collection(this.auditColl);
            const result = await coll.insertOne({
                collName     : collName,
                collDocuments: collDocuments,
                logType      : "read",
                logBy        : userId,
                logAt        : new Date(),
            });

            if (result) {
                return getResMessage("success", {
                    value: result,
                });
            } else {
                return getResMessage("insertError");
            }
        } catch (error) {
            console.error("Error inserting read/search-audit record(s): ", error);
            return getResMessage("insertError", {
                message: "Error inserting read/search-audit record(s):" + error.message,
            });
        }
    }

    async deleteLog(collName: string, collDocuments: any, userId: string): Promise<ResponseMessage> {
        const dbCheck = checkDb(this.dbConnect);
        if (!dbCheck) {
            return dbCheck;
        }

        // Check/validate the attributes / parameters
        let errorMessage = "";
        if (!collName) {
            errorMessage = errorMessage ? errorMessage + " | Table or Collection name is required." :
                "Table or Collection name is required.";
        }
        if (!userId) {
            errorMessage = errorMessage ? errorMessage + " | userId is required." :
                "userId is required.";
        }
        if (!collDocuments) {
            errorMessage = errorMessage ? errorMessage + " | Deleted record(s) information is required." :
                "Deleted record(s) information is required.";
        }
        if (errorMessage) {
            return getResMessage("insertError", {
                message: errorMessage,
            });
        }

        try {
            // insert audit record
            const coll = this.dbConnect.collection(this.auditColl);
            const result = await coll.insertOne({
                collName     : collName,
                collDocuments: collDocuments,
                logType      : "remove",
                logBy        : userId,
                logAt        : new Date(),
            });

            if (result) {
                return getResMessage("success", {
                    value: result,
                });
            } else {
                return getResMessage("insertError");
            }
        } catch (error) {
            console.log("Error saving delete-audit record(s): ", error);
            return getResMessage("insertError", {
                message: "Error inserting delete-audit record(s):" + error.message,
            });
        }
    }

    async loginLog(collDocuments: any, userId: string = ""): Promise<ResponseMessage> {
        const dbCheck = checkDb(this.dbConnect);
        if (!dbCheck) {
            return dbCheck;
        }
        // validate params/values
        let errorMessage = "";
        if (!collDocuments) {
            errorMessage = errorMessage + " | Login information is required."
        }
        if (errorMessage) {
            return getResMessage("insertError", {
                message: errorMessage,
            });
        }

        try {
            // insert audit record
            const coll = this.dbConnect.collection(this.auditColl);
            const result = await coll.insertOne({
                collDocuments: collDocuments,
                logType      : "login",
                logBy        : userId,
                logAt        : new Date(),
            });

            if (result) {
                return getResMessage("success", {
                    value: result,
                });
            } else {
                return getResMessage("insertError");
            }
        } catch (error) {
            console.log("Error inserting login-audit record(s): ", error);
            return getResMessage("insertError", {
                message: "Error inserting login-audit record(s):" + error.message,
            });
        }
    }

    async logoutLog(collDocuments: any, userId: string = ""): Promise<ResponseMessage> {
        const dbCheck = checkDb(this.dbConnect);
        if (!dbCheck) {
            return dbCheck;
        }

        // validate params/values
        let errorMessage = "";
        if (!userId) {
            errorMessage = errorMessage + " | userId is required."
        }
        if (!collDocuments || collDocuments === "") {
            errorMessage = errorMessage + " | Logout information is required."
        }
        if (errorMessage) {
            return getResMessage("insertError", {
                message: errorMessage,
            });
        }

        try {
            // insert audit record
            const coll = this.dbConnect.collection(this.auditColl);
            const result = await coll.insertOne({
                collDocuments: collDocuments,
                logType      : "logout",
                logBy        : userId,
                logAt        : new Date(),
            });

            if (result) {
                return getResMessage("success", {
                    value: result,
                });
            } else {
                return getResMessage("insertError");
            }
        } catch (error) {
            console.log("Error inserting logout-audit record(s): ", error);
            return getResMessage("insertError", {
                value: error,
            });
        }
    }

    async auditLog(logType: string, userId: string = "", options?: AuditLogOptionsType) {
        const dbCheck = checkDb(this.dbConnect);
        if (!dbCheck) {
            return dbCheck;
        }

        // Check/validate the attributes / parameters
        let collName = "",
            collDocuments = null,
            newCollDocuments = null,
            errorMessage = "",
            queryParams = {};

        logType = logType.toLowerCase();

        switch (logType) {
            case "create":
            case AuditLogTypes.CREATE:
                collName = options && options.collName ? options.collName : "";
                collDocuments = options && options.collDocuments ? options.collDocuments : null;
                // validate params/values
                if (!collName) {
                    errorMessage = errorMessage ? errorMessage + " | Table or Collection name is required." :
                        "Table or Collection name is required.";
                }
                if (!userId) {
                    errorMessage = errorMessage ? errorMessage + " | userId is required." :
                        "userId is required.";
                }
                if (!collDocuments) {
                    errorMessage = errorMessage ? errorMessage + " | Created record(s) information is required." :
                        "Created record(s) information is required.";
                }
                if (errorMessage) {
                    return getResMessage("insertError", {
                        message: errorMessage,
                    });
                }
                queryParams = {
                    collName     : collName,
                    collDocuments: collDocuments,
                    logType      : logType,
                    logBy        : userId,
                };
                break;
            case "update":
            case AuditLogTypes.UPDATE:
                collName = options && options.collName ? options.collName : "";
                collDocuments = options && options.collDocuments ? options.collDocuments : null;
                newCollDocuments = options && options.newCollDocuments ? options.newCollDocuments : null; // object or array

                // validate params/values
                if (!collName) {
                    errorMessage = errorMessage ? errorMessage + " | Table or Collection name is required." :
                        "Table or Collection name is required.";
                }
                if (!userId) {
                    errorMessage = errorMessage ? errorMessage + " | userId is required." :
                        "userId is required.";
                }
                if (!collDocuments) {
                    errorMessage = errorMessage ? errorMessage + " | Current record(s) information is required." :
                        "Current record(s) information is required.";
                }
                if (!newCollDocuments) {
                    errorMessage = errorMessage ? errorMessage + " | Updated record(s) information is required." :
                        "Updated record(s) information is required.";
                }
                if (errorMessage) {
                    return getResMessage("insertError", {
                        message: errorMessage,
                    });
                }

                queryParams = {
                    collName        : collName,
                    collDocuments   : collDocuments,
                    newCollDocuments: newCollDocuments,
                    logType         : logType,
                    logBy           : userId,
                };
                break;
            case "remove":
            case "delete":
            case AuditLogTypes.DELETE:
            case AuditLogTypes.REMOVE:
                collName = options && options.collName ? options.collName : "";
                collDocuments = options && options.collDocuments ? options.collDocuments : null;

                // Check/validate the attributes / parameters
                if (!collName) {
                    errorMessage = errorMessage ? errorMessage + " | Table or Collection name is required." :
                        "Table or Collection name is required.";
                }
                if (!userId) {
                    errorMessage = errorMessage ? errorMessage + " | userId is required." :
                        "userId is required.";
                }
                if (!collDocuments) {
                    errorMessage = errorMessage ? errorMessage + " | Deleted record(s) information is required." :
                        "Deleted record(s) information is required.";
                }
                if (errorMessage) {
                    return getResMessage("insertError", {
                        message: errorMessage,
                    });
                }

                queryParams = {
                    collName     : collName,
                    collDocuments: collDocuments,
                    logType      : logType,
                    logBy        : userId,
                };
                break;
            case "read":
            case AuditLogTypes.GET:
            case AuditLogTypes.READ:
                collName = options && options.collName ? options.collName : "";
                collDocuments = options && options.collDocuments ? options.collDocuments : null;

                // validate params/values
                if (!collName) {
                    errorMessage = errorMessage ? errorMessage + " | Table or Collection name is required." :
                        "Table or Collection name is required.";
                }
                if (!collDocuments) {
                    errorMessage = errorMessage ? errorMessage + " | Search keywords or Read record(s) information is required." :
                        "Search keywords or Read record(s) information is required.";
                }
                if (errorMessage) {
                    return getResMessage("insertError", {
                        message: errorMessage,
                    });
                }

                queryParams = {
                    collName     : collName,
                    collDocuments: collDocuments,
                    logType      : logType,
                    logBy        : userId,
                };
                break;
            case "login":
            case AuditLogTypes.LOGIN:
                collDocuments = options && options.collDocuments ? options.collDocuments : null;

                // validate params/values
                if (!collDocuments) {
                    errorMessage = errorMessage + " | Login information is required."
                }
                if (errorMessage) {
                    return getResMessage("insertError", {
                        message: errorMessage,
                    });
                }

                queryParams = {
                    collDocuments: collDocuments,
                    logType      : logType,
                    logBy        : userId,
                };
                break;
            case "logout":
            case AuditLogTypes.LOGOUT:
                collDocuments = options && options.collDocuments ? options.collDocuments : null;

                // validate params/values
                if (!userId) {
                    errorMessage = errorMessage + " | userId is required."
                }
                if (!collDocuments || collDocuments === "") {
                    errorMessage = errorMessage + " | Logout information is required."
                }
                if (errorMessage) {
                    return getResMessage("insertError", {
                        message: errorMessage,
                    });
                }
                queryParams = {
                    collDocuments: collDocuments,
                    logType      : logType,
                    logBy        : userId,
                };
                break;
            default:
                return getResMessage("insertError", {
                    message: "Unknown log type and/or incomplete log information",
                });
        }

        try {
            // insert audit record
            const coll = this.dbConnect.collection(this.auditColl);
            const result = await coll.insertOne({
                ...queryParams,
                ...{logAt: new Date()},
            });

            if (result) {
                return getResMessage("success", {
                    value: result,
                });
            } else {
                return getResMessage("insertError");
            }
        } catch (error) {
            console.log("Error saving audit-log record(s): ", error);
            return getResMessage("insertError", {
                message: "Error inserting audit-log record(s):" + error.message,
            });
        }
    }
}

function newAuditLogMongo(auditDb: MongoDbConnectType, options?: AuditLogOptionsType) {
    return new AuditLogMongo(auditDb, options);
}

export { AuditLogMongo, newAuditLogMongo };
